### DISCUSSION TITLE
KlipperScreen custom panels/plugins

### DISCUSSION THREAD

**Discussion by Jomik** [KlipperScreen/KlipperScreen#1393] [Ideas]
I am a bit surprised that this is not a thing.
How do we add a panel / modify the "base" panel?
Is it only by forking and distributing a specialised version?
Thinking of Happy Hare as an example.

I think it would be great if it was possible to add your own panels in a way, or even install community made ones.

We have a simple MMU coming up that would be awesome to support through KlipperScreen, but it probably doesn't fit in the core repo, as this would become bloated with support for too many things.
We should be able to look for modules in a special location or define them in a configuration file?

**Comment by alfrix** [comment]
> How do we add a panel / modify the "base" panel?
> Is it only by forking and distributing a specialised version?
> Thinking of Happy Hare as an example.

hh seems to be a very involved modification, forking was the choice of the developer of the project.

> I think it would be great if it was possible to add your own panels in a way, or even install community made ones.

it's possible to create a panel, there is an example panel that is empty in the panels folder, but requires quite a bit of knowledge of python and pygobject.

there is also the less complex approach of defining menus with gcode

you can add panels or menu entries via config (klipperscreen.conf) as described in the docs, it can even override the default config

i don't have a problem with giving you a hand if you want, i recommend that before you do any coding, you think about what should be shown and how, ideally a rough sketch

in the case of going without a fork i would guess that you would create a repo with a config file and panels, and inject that with an   [include] section in the user config

**Reply by phnxdrgn536** [reply]
is there a way to change the format of the time readouts like elapsed time on the print status screen for "hours:hh minutes:mm" to something along the lines of hh:mm:ss??
